Forta-Bot-Challenge-4/COMP01/src/thresholdCache/thresholdCache.ts

```diff
@@ -0,0 +1,108 @@
+import { LRUCache } from "lru-cache";
+import { ethers } from "ethers";
+
+import { BigNumber } from "@ethersproject/bignumber";
+
+export let THRESHOLD: any = ethers.BigNumber.from("10");
+
+export const amountCache: LRUCache<string, { timestamp: number; amount: any }> = new LRUCache({
+  max: 1000,
+});
+
+export const amountOverThreshold = async (userAddress: string, amount: any, timestamp: number): Promise<number> => {
+  const bigAmount = BigNumber.from(amount);
+  const bigThreshold = BigNumber.from(THRESHOLD);
+
+  // Check if there is an entry in the cache for the current userAddress
+  if (amountCache.has(userAddress)) {
+    const cachedData = amountCache.get(userAddress) || { timestamp: 0, amount: "0" };
+    const totalAmountWithinDay = BigNumber.from(cachedData.amount).add(bigAmount);
+
+    // Update the cache with the provided timestamp and accumulated amount
+    amountCache.set(userAddress, {
+      timestamp: timestamp,
+      amount: totalAmountWithinDay.toString(),
+    });
+
+    // Check if the total amount within the last day is over the threshold
+    if (totalAmountWithinDay.gt(bigThreshold)) {
+      return totalAmountWithinDay.sub(bigThreshold).toNumber();
+    }
+  } else {
+    // If there is no entry in the cache, update the cache with the provided timestamp and amount
+    amountCache.set(userAddress, {
+      timestamp: timestamp,
+      amount: bigAmount.toString(),
+    });
+
+    // Check if the current amount is greater than the threshold
+    if (bigAmount.gt(bigThreshold)) {
+      return bigAmount.sub(bigThreshold).toNumber();
+    }
+  }
+
+  // Return 0 since the total amount within the last day is not over the threshold
+  return 0;
+};
+
+// Function to update the threshold as 20% above the average of the amounts stored in the cache
+export const calculateNewThreshold = (): void => {
+  let totalAmount = BigNumber.from("0");
+  let count = 0;
+
+  // Get the generator object
+  const cacheValuesGenerator = amountCache.values();
+
+  // Convert the generator into an array using Array.from
+  const cacheValues = Array.from(cacheValuesGenerator).filter((value) => value !== undefined) as {
+    timestamp: number;
+    amount: any;
+  }[];
+
+  // Iterate over the array
+  for (let i = 0; i < cacheValues.length; i++) {
+    const data = cacheValues[i];
+    totalAmount = totalAmount.add(BigNumber.from(data.amount));
+    count++;
+  }
+
+  if (count === 0) {
+    // Do not update the threshold if there are no entries in the cache
+    return;
+  }
+
+  // Calculate the average
+  const average = totalAmount.div(count);
+
+  // Calculate the new threshold as 20% above the average
+  const newThreshold = average.add(average.mul(20).div(100));
+
+  // Update the global threshold
+  THRESHOLD = newThreshold;
+};
+
+// Function to clear the cache
+export const clearCache = () => {
+  amountCache.clear();
+};
+// Function to clear the cache every 24 hours
+export const clearCachePeriodically = () => {
```

Comment:
I understand what this function is attempting to do, _but_ I would advise against it because it'll be deceiving in how it actually affects your bot.

This will clear your cache at a fixed time every 24 hours. However, from my understanding of your implementation, you will be treating each "supplier" individually with their own 24 hour time window.

Let's say you cache clears every midnight on the dot, for simplicity. Then you have someone supply at 9pm. Then second supply the next day at 9am. Your cache would've clear at midnight, so the second supply wouldn't have been added to the first. What if that second supply wouldn't caused it to exceed the threshold, your bot would've missed it because it clears at the same time.

Therefore, the 24 hour time window for each supply should begin at the time of the first supply and close 24 hours after. In the example I gave, the window would've closed at 9pm the next, and thus would've caught the supply at 9am.

So best to track the deposits per user rather than a blanket time for all users.